packages/babel-parser/src/parser/comments.js

```diff
@@ -1,288 +1,214 @@
 // @flow
 
+/*:: declare var invariant; */
+
+import BaseParser from "./base";
+import type { Comment, Node } from "../types";
+import * as charCodes from "charcodes";
+
 /**
- * Based on the comment attachment algorithm used in espree and estraverse.
- *
- * Redistribution and use in source and binary forms, with or without
- * modification, are permitted provided that the following conditions are met:
- *
- * * Redistributions of source code must retain the above copyright
- *   notice, this list of conditions and the following disclaimer.
- * * Redistributions in binary form must reproduce the above copyright
- *   notice, this list of conditions and the following disclaimer in the
- *   documentation and/or other materials provided with the distribution.
+ * A whitespace token containing comments
+ * @typedef CommentWhitespace
+ * @type {object}
+ * @property {number} start - the start of the whitespace token.
+ * @property {number} end - the end of the whitespace token.
+ * @property {Array<Comment>} comments - the containing comments
+ * @property {Node | null} leadingNode - the immediately preceding AST node of the whitespace token
+ * @property {Node | null} trailingNode - the immediately following AST node of the whitespace token
+ * @property {Node | null} containerNode - the outermost AST node containing the whitespace
```

Comment:
I think in the future we can change the behaviour to have `innerComments` attached to the _innermost_ AST node. So in the example `foo (/* 2 */)`, `/* 2 */` will the an `innerComments` of call expression. The idea is to provide the generator more information about how to insert `innerComments`.

Or should we change in this PR now?